### PR TITLE
[Merged by Bors] - feat: `lake exe graph`, as replacement for `leanproject import-graph`

### PR DESCRIPTION
**Pull request by kim-em**
Example usage:
```
lake exe graph --from Mathlib.Topology.Constructions --to Mathlib/Topology/Homotopy/Contractible.lean --reduce import_graph.pdf
```
producing

- [import_graph.pdf](https://github.com/leanprover-community/mathlib4/files/11877278/import_graph.pdf)

Current features:
* supports `--to` and `--from` to select a slice of the full import graph
* accepts either module names or file names
* supports `--reduce` to perform transitive reduction
* output to `.dot` files, or if you have graphviz installed output to any supported format (e.g. `.pdf` as above).

Still missing:
* `--show-unused` (although all the machinery is now in place to do this)
* `--exclude-tactics`

---

- [x] depends on: #5441

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#5441
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#5441~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: lakefile.lean [hunk 1]
```diff
@@ -48,3 +48,9 @@ lean_lib Archive where
 
 lean_lib Counterexamples where
   roots := #[`Counterexamples]
+
+lean_lib ImportGraph where
+  roots := #[`ImporGraph]
```
Comment (sgouezel) [inline lakefile.lean:53]:
```suggestion
  roots := #[`ImportGraph]
```

File: Mathlib/Lean/IO/Process.lean [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+namespace IO.Process
+
+open System (FilePath)
+
+/--
+Pipe `input` into stdin of the spawned process,
+then return `(exitCode, stdout, stdErr)` upon completion.
+-/
+-- TODO We could reduce some code duplication by centralising some functions like this,
+-- which are used here, in `cahce`, and in https://github.com/leanprover-community/llm.
```
Comment (alexjbest) [inline Mathlib/Lean/IO/Process.lean:20]:
```suggestion
-- which are used here, in `cache`, and in https://github.com/leanprover-community/llm.
```

**Comment by alexjbest** [comment]
Can such things (e.g. this and cache) live in a folder like Util or Scripts or some such? Or do they really need to be separate modules within the root folder, and not just a submodule inside Mathlib? I fear we will end up with it being harder for new (and old) users who just open the github page to find where all the interesting stuff is if we get more scripts at top level.

**Comment by semorrison** [comment]
@alexjbest I investigated moving `Cache` and `ImportGraph` under a common `Util` directory, but it seems `lake` can't cope with this. I propose we don't block this PR on this issue, and I'll follow up separately.

**Comment by alexjbest** [comment]
Ok, thanks for looking into it.
Lgtm then

maintainer merge

**Pull request closed without merging**